applications/RomApplication/custom_strategies/custom_schemes/modal_derivative_scheme.hpp

```diff
@@ -0,0 +1,655 @@
+
+        // Compute ElementMatrixDerivative
+        noalias(rElementMatrixDerivative) = (element_matrix_p_perturbed - element_matrix) / mFiniteDifferenceStepSize;
+
+        // Unlock element nodes
+        this->UnlockElementNodes(rElement);
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This function performs central differencing on the element LHS matrix
+     * @details 
+     * @param rElementMatrixDerivative The element to unlock nodes
+     * @param basisIndex The index of the basis for perturbation
+     * @param rElement The element to compute finite differencing
+     * @param rCurrentProcessInfo The current process info instance
+     */
+    void CentralDifferencingWithBasis(
+         LocalSystemMatrixType& rElementMatrixDerivative,
+         const std::size_t basisIndex,
+         Element& rElement,
+         const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+
+        // Lock element nodes for OMP parallelism
+        this->LockElementNodes(rElement);
+
+        const auto matrix_size = rElementMatrixDerivative.size1();
+
+        // Positive perturbation
+        LocalSystemMatrixType element_matrix_p_perturbed;
+        element_matrix_p_perturbed.resize(matrix_size, matrix_size, false);
+        this->PerturbElementWithBasis(1.0, basisIndex, rElement, rCurrentProcessInfo);
+        rElement.CalculateLeftHandSide(element_matrix_p_perturbed, rCurrentProcessInfo);
+
+        // Negative perturbation
+        LocalSystemMatrixType element_matrix_n_perturbed;
+        element_matrix_n_perturbed.resize(matrix_size, matrix_size, false);
+        this->PerturbElementWithBasis(-2.0, basisIndex, rElement, rCurrentProcessInfo);
+        rElement.CalculateLeftHandSide(element_matrix_n_perturbed, rCurrentProcessInfo);
+
+        // Reset perturbation
+        this->PerturbElementWithBasis(1.0, basisIndex, rElement, rCurrentProcessInfo);
+
+        // Compute LHS derivative
+        noalias(rElementMatrixDerivative) = (element_matrix_p_perturbed - element_matrix_n_perturbed) / (2.0 * mFiniteDifferenceStepSize);
+
+        this->UnlockElementNodes(rElement);
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This function perturbs the element with the given basisIndex
+     * @details 
+     * @param Step The step direction and size
+     * @param basisIndex The index of the basis for perturbation
+     * @param rElement The element to perturb
+     * @param rCurrentProcessInfo The current process info instance
+     */
+    // This function perturbs the element with the vector with given eigenvector index basisIndex
+    void PerturbElementWithBasis(
+        const double Step,
+        const std::size_t basisIndex,
+        Element& rElement,
+        const ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY
+
+        // Initialize is necessary for updating the section properties of shell elements
+        rElement.InitializeNonLinearIteration(rCurrentProcessInfo);
```

Comment:
hm you sure abt this?
I would do this globally and not hidden here

Reply:
The element internal stuff should be updated at each FD step. Another option is doing the FD from the strategy level with the complete system matrices, but then I will have to store one more system matrix for forward FD and two more system matrices for central FD. It is not memory efficient. So, I dont know where else it could be done

Reply:
then I would at least remove the comment regarding the shells
Otherwise seems like a hack

Reply:
It is almost only for the currently implemented shells actually. So it is like a hack but not a hack. :)
I thought it would be good to have it there to make sure it doesnt go away in a round of clean up, unless the shell implementation changes.